### PR TITLE
Altered accounts integration with drivers

### PR DESCRIPTION
**Pull request by bogdan-rosianu**
integrated outport drivers with support for altered accounts

### CODE REVIEW

File: factory/processComponents.go [hunk 1]
```diff
@@ -478,6 +482,18 @@ func (pcf *processComponentsFactory) Create() (*processComponents, error) {
 		return nil, err
 	}
 
+	pcf.esdtNftStorage, err = vmcommonBuiltInFunctions.NewESDTDataStorage(vmcommonBuiltInFunctions.ArgsNewESDTDataStorage{
```
Comment (iulianpascalau) [inline factory/processComponents.go:485]: could have had an arg variable here, it is harder to read like this

File: process/block/alteredaccounts/alteredAccountsProvider.go [hunk 2]
```diff
@@ -151,17 +158,13 @@ func (aap *alteredAccountsProvider) addTokensDataForMarkedAccount(
...
-	esdtTokenBytes, err := userAccount.RetrieveValueFromDataTrieTracker(storageKey)
-	if err != nil {
-		log.Warn("cannot get ESDT data for address",
-			"address", encodedAddress,
-			"tokenKey", tokenKey,
-			"error", err)
+	userAccountVmCommon, ok := userAccount.(vmcommon.UserAccountHandler)
+	if !ok {
+		log.Warn("state.UserAccountHandler cannot be cast to vmcommon.UserAccountHandler", "address", aap.addressConverter.Encode(userAccount.GetOwnerAddress()))
 		return nil
```
Comment (iulianpascalau) [inline process/block/alteredaccounts/alteredAccountsProvider.go:164]: why not returning error here?

File: factory/processComponents.go (hunk 1 above)
Comment (bogdan-rosianu) [reply factory/processComponents.go:485]: done

File: process/block/alteredaccounts/alteredAccountsProvider.go (hunk 2 above)
Comment (bogdan-rosianu) [reply process/block/alteredaccounts/alteredAccountsProvider.go:164]: you're right. it's a programming error. returned error + added unit test

Comment (iulianpascalau) [reply process/block/alteredaccounts/alteredAccountsProvider.go]: :hat-off2:

File: process/block/alteredaccounts/alteredAccountsProvider.go [hunk 3]
```diff
@@ -125,7 +125,7 @@ func (aap *alteredAccountsProvider) processMarkedAccountData(
 	userAccount, ok := account.(state.UserAccountHandler)
 	if !ok {
 		log.Warn("cannot cast AccountHandler to UserAccountHandler", "address", encodedAddress)
```
Comment (iulianpascalau) [inline process/block/alteredaccounts/alteredAccountsProvider.go:127]: we are recording the error as a WARN in the block processors. Suggestion to remove these WARNs and return a wrapped error. Something like:
```GO
return fmt.Errorf("%w while casting to UserAccountsHandler for address %s", errCannotCastToUserAccountHandler, encodedAddress)
```
Valid for this line, L119, L140, L163

Comment (bogdan-rosianu) [reply process/block/alteredaccounts/alteredAccountsProvider.go]: done

**Review by iulianpascalau — DISMISSED** [review]

**Review by gabi-vuls — DISMISSED** [review]
System test passed.

File: process/block/alteredaccounts/tokensProcessor.go [hunk 4]
```diff
@@ -13,6 +13,13 @@ const (
 	idxTokenIDInTopics         = 0
 	idxTokenNonceInTopics      = 1
 	idxReceiverAddressInTopics = 3
+	issueFungibleESDTFunc      = "issue"
```
Comment (iulianpascalau) [inline process/block/alteredaccounts/tokensProcessor.go:16]: duplicated declarations with the elrond-go-core/core package?

Comment (bogdan-rosianu) [reply process/block/alteredaccounts/tokensProcessor.go:16]: I cannot find them declared elsewhere

**Review by popenta — APPROVED** [review]
@@                            Log scanner                                  @@
##                            altered-accounts-integration                 ##
================================================================================
- Known Warnings              666
- New Warnings                3
- Known Errors                4
- New Errors                  2
- Panics                      0
================================================================================

File: process/block/alteredaccounts/tokensProcessor.go [hunk 5]
```diff
@@ -113,7 +129,8 @@ func (tp *tokensProcessor) extractEsdtData(
 		return err
 	}
 
-	if len(topics) > idxReceiverAddressInTopics {
+	eventShouldContainReceiverAddress := string(event.GetIdentifier()) != core.BuiltInFunctionESDTNFTCreate
```
Comment (miiu96) [inline process/block/alteredaccounts/tokensProcessor.go:132]: I think events with identifier ESDTNFTTransfer, ESDTTransfer, MultiTransfer And Wipe contains receiver address on the topics[3]

**Review by miiu96 — CHANGES_REQUESTED** [review]

File: factory/export_test.go [hunk 6]
```diff
@@ -100,5 +101,5 @@ func SetShardCoordinator(shardCoordinator sharding.Coordinator, holder Bootstrap
 
 // IndexGenesisBlocks -
 func (pcf *processComponentsFactory) IndexGenesisBlocks(genesisBlocks map[uint32]data.HeaderHandler, indexingData map[uint32]*genesis.IndexingData) error {
-	return pcf.indexGenesisBlocks(genesisBlocks, indexingData)
+	return pcf.indexGenesisBlocks(genesisBlocks, indexingData, map[string]*indexer.AlteredAccount{})
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply factory/export_test.go:104]: you commented on `export_test.go`. in production, real accounts are forwarded

File: process/block/alteredaccounts/tokensProcessor.go (hunk 5 above)
Comment (bogdan-rosianu) [reply process/block/alteredaccounts/tokensProcessor.go:132]: true. refactored

File: go.mod [hunk 7]
```diff
@@ -9,12 +9,12 @@ require (
...
 	github.com/ElrondNetwork/covalent-indexer-go v1.0.7-0.20220211153054-65a2a0e42ae2
 	github.com/ElrondNetwork/elastic-indexer-go v1.1.41-0.20220406120732-c0a926362226
-	github.com/ElrondNetwork/elrond-go-core v1.1.14
+	github.com/ElrondNetwork/elrond-go-core v1.1.15
 	github.com/ElrondNetwork/elrond-go-crypto v1.0.1
 	github.com/ElrondNetwork/elrond-go-logger v1.0.7
 	github.com/ElrondNetwork/elrond-vm-common v1.3.2
 	github.com/ElrondNetwork/go-libp2p-pubsub v0.5.5-rc2
-	github.com/ElrondNetwork/notifier-go v1.0.4-0.20220201135024-c6fdc94cb5a1
+	github.com/ElrondNetwork/notifier-go v1.1.1-0.20220427110312-750b7f81d15d
```
Comment (iulianpascalau) [inline go.mod:17]: proper release?

File: process/block/alteredaccounts/tokensProcessor.go [hunk 8]
```diff
@@ -106,14 +122,20 @@ func (tp *tokensProcessor) extractEsdtData(
...
 
-	if len(topics) > idxReceiverAddressInTopics {
+	// in case of esdt transfer, nft transfer, wipe or multi esdt transfers, the 3rd index of the topics contains the destination address
+	identifier := string(event.GetIdentifier())
+	eventShouldContainReceiverAddress := identifier == core.BuiltInFunctionESDTTransfer ||
+		identifier == core.BuiltInFunctionESDTNFTTransfer ||
+		identifier == core.BuiltInFunctionESDTWipe ||
+		identifier == core.BuiltInFunctionMultiESDTNFTTransfer
+
+	if eventShouldContainReceiverAddress && len(topics) > idxReceiverAddressInTopics {
```
Comment (miiu96) [inline process/block/alteredaccounts/tokensProcessor.go:138]: :+1:

**Review by miiu96 — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed.

**Pull request merged**